server/src/updaters/relationship-updaters.js

```diff
@@ -0,0 +1,158 @@
+// @flow
+
+import type { Viewer } from '../session/viewer';
+
+import invariant from 'invariant';
+import _groupBy from 'lodash/fp/groupBy';
+import _keyBy from 'lodash/fp/keyBy';
+
+import {
+  type RelationshipRequest,
+  knowOfFriendsStatus,
+  requestsBlocksStatus,
+} from 'lib/types/relationship-types';
+import { ServerError } from 'lib/utils/errors';
+
+import { fetchUserInfos } from '../fetchers/user-fetchers';
+
+import { dbQuery, SQL } from '../database';
+
+type Errors = {
+  invalid_user?: string[],
+};
+
+async function updateRelationship(
+  viewer: Viewer,
+  request: RelationshipRequest,
+) {
+  const { action } = request;
+
+  if (!viewer.loggedIn) {
+    throw new ServerError('not_logged_in');
+  }
+
+  const uniqueUserIDs = [...new Set(request.userIDs)];
+  const users = await fetchUserInfos(uniqueUserIDs);
+
+  const errors: Errors = {};
+  const userIDs: string[] = [];
+  for (let userID of uniqueUserIDs) {
+    if (userID === viewer.userID || !users[userID].username) {
+      const acc = errors.invalid_user || [];
+      errors.invalid_user = [...acc, userID];
+    } else {
+      userIDs.push(userID);
+    }
+  }
+
+  if (action === 'friend_request') {
+    const requestsOrBlockedSelectQuery = SQL`
+      SELECT user1, user2, status from friend_requests_blocks 
+      WHERE (
+        (user1 = ${viewer.userID} AND user2 IN (${userIDs}))
+        OR (user1 IN (${userIDs}) AND user2 = ${viewer.userID}) 
+        AND status = ${requestsBlocksStatus.PENDING_FRIEND}
+      ) 
+      OR user1 = ${viewer.userID} AND status = ${requestsBlocksStatus.BLOCKED}
+    `;
+
+    const friendsOrBlockedMeSelectQuery = SQL`
+      SELECT user1, user2, status from know_of_friends 
+      WHERE (user1 = ${viewer.userID} OR user2 = ${viewer.userID}) AND status = ${knowOfFriendsStatus.FRIEND}
+      UNION SELECT user1, user2, status from friend_requests_blocks 
+      WHERE user2 = ${viewer.userID} AND status = ${requestsBlocksStatus.BLOCKED}
+    `;
+
+    const [
+      [requestsOrBlockedResults],
+      [friendsOrBlockedMeResult],
+    ] = await Promise.all([
+      dbQuery(requestsOrBlockedSelectQuery),
+      dbQuery(friendsOrBlockedMeSelectQuery),
+    ]);
+
+    const knowOfFriendInsertRows = [];
+    const requestBlockInsertRows = [];
+    const requestBlockDeleteIDs = [];
+    const restUserIds = _keyBy(id => id, userIDs);
+    for (const relationship of requestsOrBlockedResults) {
```

Comment:
I didn't realize you could use `const` in `for-of` and `for-in` loops! Had to do some Googling about this, learned something today :) Apparently it doesn't work in the classic `for` loop though